File: Android/app/src/main/java/com/lunesu/pengchauferry/FerryPier.kt
package com.lunesu.pengchauferry

enum class FerryPier {
    Central,
    PengChau,
    TrappistMonastery,
    DiscoveryBay,
    MuiWo,
    CheungChau,
    ChiMaWan,
    HeiLingChau;

    val coordinate get() = COORDS.getValue(this)

    companion object {
        internal val ENUMS = values()

        val COORDS = mapOf(
            Central to Coordinate(22.2871, 114.1602),
            PengChau to Coordinate(22.2846, 114.0379),
            TrappistMonastery to Coordinate(22.2819, 114.0225),
            DiscoveryBay to Coordinate(22.2937, 114.0210),
            MuiWo to Coordinate(22.2647, 114.0019),
            CheungChau to Coordinate(22.2084, 114.0287),
            ChiMaWan to Coordinate(22.2393, 113.9995),
            HeiLingChau to Coordinate(22.2579, 114.0280)
        )

        fun findNearest(latitude: Double, longitude: Double, piers: Array<FerryPier> = ENUMS): FerryPier? {
            return piers.minBy { it.coordinate.distance(latitude, longitude) }
        }
    }

}

File: iOS/PengChauFerry/FerryPier.swift
//
//  FerryPier.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 7/3/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import Foundation

public enum FerryPier: String, Codable {
    case Central
    case PengChau
    case TrappistMonastery
    case DiscoveryBay
    case MuiWo
    case CheungChau
    case ChiMaWan
    case HeiLingChau

    static let ENUMS = [Central, PengChau, TrappistMonastery, DiscoveryBay, MuiWo, CheungChau, ChiMaWan, HeiLingChau]

    static let COORDS: [FerryPier: Coordinate] = [
        .Central: Coordinate(latitude: 22.2871, longitude: 114.1602),
        .PengChau: Coordinate(latitude: 22.2846, longitude: 114.0379),
        .TrappistMonastery: Coordinate(latitude: 22.2819, longitude: 114.0225),
        .DiscoveryBay: Coordinate(latitude: 22.2937, longitude: 114.0210),
        .MuiWo: Coordinate(latitude: 22.2647, longitude: 114.0019),
        .CheungChau: Coordinate(latitude: 22.2084, longitude: 114.0287),
        .ChiMaWan: Coordinate(latitude: 22.2393, longitude: 113.9995),
        .HeiLingChau: Coordinate(latitude: 22.2579, longitude: 114.0280)
    ]

    var coordinate: Coordinate {
        get { FerryPier.COORDS[self]! }
    }

    static func findNearest(latitude: Double, longitude: Double, piers: [FerryPier] = ENUMS) -> FerryPier? {
        let dists = piers.map { (key: $0, value: COORDS[$0]!.distance(lat: latitude, long: longitude)) }
        return dists.min { $0.value < $1.value }?.key
    }

    static func valueOf(_ s: String) -> FerryPier? {
        FerryPier(rawValue: s)
    }

    func toString() -> String {
        return rawValue
    }
}

extension FerryPier: Identifiable {
    public var id: String {
        get { self.rawValue }
    }
}
